mullvad-relay-selector/src/relay_selector/matcher.rs

```diff
@@ -153,6 +141,44 @@ fn filter_on_obfuscation(
     }
 }
 
+// When core privacy features are enabled, only a subset of WireGuard relays will be included
+fn filter_on_core_privacy_features(query: &WireguardRelayQuery, relay: &Relay) -> bool {
+    filter_on_daita(&query.daita, relay)
+        && filter_on_multihop(query.multihop(), relay)
+        && filter_on_quantum_resistance(query.quantum_resistant, relay)
+}
+
+/// Filter out relays that do not support WireGuard multihop.
+fn filter_on_multihop(enabled: bool, relay: &Relay) -> bool {
+    // When multihop is enabled, discard all non-wireguard relays
+    !enabled || matches!(relay.endpoint_data, RelayEndpointData::Wireguard(_))
+}
+
+/// Filter out relays that do not support quantum resistance.
+fn filter_on_quantum_resistance(state: QuantumResistantState, relay: &Relay) -> bool {
+    // When PQ is enabled, discard all non-wireguard relays
+    if state == QuantumResistantState::On {
+        matches!(relay.endpoint_data, RelayEndpointData::Wireguard(_))
+    } else {
+        true
+    }
+}
+
+/// Returns whether `relay` satisfy the daita constraint posed by `filter`.
+pub fn filter_on_daita(filter: &Constraint<bool>, relay: &Relay) -> bool {
+    match (filter, &relay.endpoint_data) {
+        // Only a subset of relays support DAITA, so filter out ones that don't.
+        (
+            Constraint::Only(true),
+            RelayEndpointData::Wireguard(WireguardRelayEndpointData { daita, .. }),
+        ) => *daita,
+        // Filter out all non-WireGuard relays
+        (Constraint::Only(true), _) => false,
+        // Keep all relays when DAITA isn't requested.
+        _ => true,
+    }
+}
```

Comment:
You can ignore this if you decide to do the above, but if you keep these, maybe it's more readable to use the same matching method for all of them?
```suggestion
// When core privacy features are enabled, only a subset of WireGuard relays will be included
fn filter_on_core_privacy_features(query: &WireguardRelayQuery, relay: &Relay) -> bool {
    filter_on_daita(&query.daita, relay)
        && filter_on_multihop(query.multihop(), relay)
        && filter_on_quantum_resistance(query.quantum_resistant, relay)
}

/// Filter out relays that do not support WireGuard multihop.
fn filter_on_multihop(enabled: bool, relay: &Relay) -> bool {
    // When multihop is enabled, discard all non-wireguard relays
    if enabled {
        matches!(relay.endpoint_data, RelayEndpointData::Wireguard(_))
    } else {
        true
    }
}

/// Filter out relays that do not support quantum resistance.
fn filter_on_quantum_resistance(state: QuantumResistantState, relay: &Relay) -> bool {
    // When PQ is enabled, discard all non-wireguard relays
    if state == QuantumResistantState::On {
        matches!(relay.endpoint_data, RelayEndpointData::Wireguard(_))
    } else {
        true
    }
}

/// Returns whether `relay` satisfy the daita constraint posed by `filter`.
pub fn filter_on_daita(filter: &Constraint<bool>, relay: &Relay) -> bool {
    if matches!(filter, Constraint::Only(true)) {
        matches!(
            &relay.endpoint_data,
            RelayEndpointData::Wireguard(WireguardRelayEndpointData { daita: true, .. })
        )
    } else {
        true
    }
}
```

Reply:
Fixed by making it impossible to construct invalid queries (as discussed).